Periodic_4_hyperbolic_triangulation_2/include/CGAL/Periodic_4_hyperbolic_Delaunay_triangulation_traits_2.h

```diff
@@ -0,0 +1,688 @@
+// Copyright (c) 2016-2018  INRIA Sophia Antipolis, INRIA Nancy (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+// You can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// Licensees holding a valid commercial license may use this file in
+// accordance with the commercial license agreement provided with the software.
+//
+// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0+
+//
+// Author(s)     :   Iordan Iordanov
+//                   Monique Teillaud
+//
+
+#ifndef CGAL_PERIODIC_4_HYPERBOLIC_DELAUNAY_TRIANGULATION_TRAITS_2_H
+#define CGAL_PERIODIC_4_HYPERBOLIC_DELAUNAY_TRIANGULATION_TRAITS_2_H
+
+#include <CGAL/license/Periodic_4_hyperbolic_triangulation_2.h>
+
+#include <CGAL/Hyperbolic_octagon_translation.h>
+#include <CGAL/Hyperbolic_Delaunay_triangulation_traits_2.h>
+
+#include <CGAL/Bbox_2.h>
+#include <CGAL/Cartesian.h>
+#include <CGAL/CORE_Expr.h>
+
+#include <boost/tuple/tuple.hpp>
+#include <boost/variant.hpp>
+
+#include <utility>
+
+namespace CGAL {
+
+template <typename K, typename Predicate_>
+class Hyperbolic_traits_with_translations_2_adaptor
+{
+  typedef K                                               Kernel;
+  typedef Predicate_                                      Predicate;
+
+  typedef typename Kernel::FT                             FT;
+  typedef typename Kernel::Point_2                        Point;
+  typedef typename Kernel::Hyperbolic_translation         Hyperbolic_translation;
+
+  // Use the construct_point_2 predicate from the kernel to convert the periodic points to Euclidean points
+  typedef typename Kernel::Construct_hyperbolic_point_2   Construct_hyperbolic_point_2;
+
+public:
+  typedef typename Predicate::result_type                 result_type;
+
+  Hyperbolic_traits_with_translations_2_adaptor() { }
+
+  result_type operator()(const Point& p0, const Point& p1,
+                         const Hyperbolic_translation& o0, const Hyperbolic_translation& o1) const
+  {
+    return Predicate()(pp(p0, o0), pp(p1, o1));
```

Comment:
The adaptor should derive from the predicate or have a const& to it, so that it doesn't default construct the predicate every time it calls it. See periodic_3 traits and their adaptors.

Reply:
Indeed, I have modified the adaptor to mimic the adaptors for Periodic_3.